feat: add proto-derived classes/enums/interfaces to the TOC

hegemonic opened the pull request:
To prevent TOC bloat, all of these items are grouped into "Classes," "Enums," and "Interfaces" sections; that seemed like a good starting point. (We could also do something fancier, like grouping items by namespace, either instead of or in addition to this. Each category can get pretty big, so my suggestion is that we create a follow-on PR to group items by namespace within each category, [like in this mockup](https://gist.github.com/hegemonic/cb3afc6dc5a3a8e2128facf7e5f88dbc)).

To see how this change affects `toc.yml`, take a look at the diff in `__snapshots__/generate-devsite.mjs.js`. I also created a gist that shows the [updated `toc.yml` for the IAM v2 client](https://gist.github.com/hegemonic/fe44505ba5965c60603110972c6a2a9e).

This PR is sort of the opposite of https://github.com/googleapis/nodejs-cloud-rad/pull/29, which fixed TOC bloat by just removing tons of links. These links are useful, though, and worth keeping.

I also threw in a bit of semi-related cleanup and fixes. Let me know if you need me to separate any of that out from this PR.

---

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com//issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea **[discussed with dsaadati@]**
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

@dansaadati FYI

hegemonic committed:
fix: update path to test spec for update-snapshots script

hegemonic committed:
fix: run YAML processors in sequence

That's how I intended for this to work. And it did work that way, but only because none of the processors were doing anything asynchronously.

When I added a processor that makes async calls, the processors started to run out of order, causing many problems.

hegemonic committed:
feat: add proto-derived classes/enums/interfaces to the TOC

To prevent TOC bloat, all of these items are grouped into "Classes," "Enums," and "Interfaces" sections; that seemed like a good starting point. (We could also do something fancier, like grouping items by namespace, either instead of or in addition to this. Each category can get pretty big, so my suggestion is that we create a follow-on PR to group items by namespace within each category, [like in this mockup](https://gist.github.com/hegemonic/cb3afc6dc5a3a8e2128facf7e5f88dbc)).

To see how this change affects `toc.yml`, take a look at the diff in `__snapshots__/generate-devsite.mjs.js`. I also created a gist that shows the [updated `toc.yml` for the IAM v2 client](https://gist.github.com/hegemonic/fe44505ba5965c60603110972c6a2a9e).

This PR is sort of the opposite of https://github.com/googleapis/nodejs-cloud-rad/pull/29, which fixed TOC bloat by just removing tons of links. These links are useful, though, and worth keeping.

I also threw in a bit of semi-related cleanup and fixes. Let me know if you need me to separate any of that out from this PR.

dansaadati commented on lib/regexp.mjs:
```diff
@@ -0,0 +1,36 @@
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+const CLASS_MATCHER = '(?:[A-Z][a-z]+)+';
```
Can we rename this to `TYPE_MATCHER`? We aren't always matching classes!

hegemonic replied:
Can you clarify this request? `CLASS_MATCHER` is meant to match classes, not anything else; as far as I can tell, that's how it's used in this PR.

dansaadati replied:
I looked at this more closely -- I agree that `CLASS_MATCHER` will match classes only so long as the library is a `GAPIC_AUTO` library type. I think it's okay to leave the name as-is, so long as we restrain using this regexp to `GAPIC_AUTO` libraries for now.

 We may need a bit more in-depth discussion of what to do about manually written libraries -- see my other comment below.

dansaadati replied:
> @dansaadati Ready for round 2! PTAL at the unresolved comments.

Left a response. Thanks again for all the work you've been putting in!

hegemonic replied:
We should be all set on non-GAPIC libraries now. If you want me to test something other than Cloud Storage, just let me know!

dansaadati commented on lib/regexp.mjs:
```diff
@@ -0,0 +1,36 @@
+// Matches `v2, someFunction, PackageClient`; captures `PackageClient`
+const EXPORTS_MATCHER = `(?:${ANY_SYMBOL_MATCHER}, )*${ANY_SYMBOL_MATCHER}`;
+// Matches `export { v2, someFunction, PackageClient };` captures `PackageClient`
+const EXPORTED_CLASS_MATCHER = 'export { ' + EXPORTS_MATCHER + ' };';
```
Same as above, can this be `EXPORTED_TYPE_REGEXP`?

hegemonic replied:
See previous question.

dansaadati commented on lib/processors/organize-toc.mjs:
```diff
@@ -0,0 +1,115 @@
+  });
+
+  Object.keys(tocItems).forEach(key => {
+    // Don't sort items with no kind, or an unrecognized kind.
```
I think we should push items with `noKind` to the bottom, so they don't appear of top-level classes and interfaces that we know are critical (i.e. the Client class). Wdyt?

dansaadati commented on lib/processors/organize-toc.mjs:
```diff
@@ -0,0 +1,115 @@
+
+const TOC_KINDS = {
+  class: 'Classes',
+  enum: 'Enums',
```
Nit: Can we show interface before enum types?

dansaadati commented on lib/processors/organize-toc.mjs:
```diff
@@ -0,0 +1,115 @@
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+import {
```
Style nit across all files: One blank line between first statement and license comment.

hegemonic commented on lib/regexp.mjs:
```diff
@@ -0,0 +1,36 @@
+// Matches `v2, someFunction, PackageClient`; captures `PackageClient`
+const EXPORTS_MATCHER = `(?:${ANY_SYMBOL_MATCHER}, )*${ANY_SYMBOL_MATCHER}`;
+// Matches `export { v2, someFunction, PackageClient };` captures `PackageClient`
+const EXPORTED_CLASS_MATCHER = 'export { ' + EXPORTS_MATCHER + ' };';
```
Agreed that we can implement option 2 in a later PR. Filed #82.

hegemonic commented on main.mjs:
I believe I made the file executable. No other changes.

hegemonic committed:
fix: address comments

hegemonic commented:
@dansaadati Ready for round 2! PTAL at the unresolved comments.

hegemonic commented on lib/regexp.mjs:
```diff
@@ -0,0 +1,37 @@
+// Matches `v2`, `someFunction`, and `PackageClient`; captures `PackageClient`
+const ANY_SYMBOL_MATCHER = `(?:${VERSION_ID_MATCHER}|${OTHER_SYMBOL_MATCHER}|(${CLASS_MATCHER}))`;
+// Matches `v2, someFunction, PackageClient`; captures `PackageClient`
+const EXPORTS_MATCHER = `(?:${ANY_SYMBOL_MATCHER}, )*${ANY_SYMBOL_MATCHER}`;
```
Ugh, I failed to consider non-autogenerated libraries. Sadness!

In light of that fact, I think it would be best to switch to an approach that involves actually parsing the code. That would make things less brittle all around; for example, that approach would be more resilient against minor changes to `@google-cloud/gapic-generator` that prevent regexps from matching. (It would also avoid the need for two separate code paths—one for autogenerated libraries, another for manually written libraries.) The only reason I didn't do it this way initially is that my somewhat hacky, regexp-centric approach seemed adequate at the time.

My hope is that I can get the necessary info from the generated `foo.api.json` file, either by loading it with [`@microsoft/api-extractor-model`](https://github.com/microsoft/rushstack/tree/main/libraries/api-extractor-model) or by examining it directly. If that doesn't pan out, I should be able to fall back on TypeScript's [Compiler API](https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API) to create an AST, then traverse it to get the required info.

Does that sound good? If not, how do you think I should proceed? Happy to chat if you like!

dansaadati replied:
I'm not particularly against having two code paths, but I agree it is more brittle than looking at the API/source code. Traversing the AST seems a bit complex -- but I like you're suggestion of using [`@microsoft/api-extractor-model`](https://github.com/microsoft/rushstack/tree/main/libraries/api-extractor-model):

We can go through each `apiMember` under the `EntryPoint`.  From there, we can treat each `apiItem` of kind `Class` or `Enum` directly under the `EntryPoint` as top-level items in the TOC. Other `Class`es `Enum`s and `Interface`s can then be put under their respective directories in the TOC.

Alternatively, we can also look at the `.d.ts` rollup file that gets generated when we run `api-extractor`, and try to grab the exports from there. I think this is a harder option, but at least we'll have an approach consistent for auto-generated and manually written libraries.

hegemonic replied:
Sounds good! I'll give `@microsoft/api-extractor-model` a shot and let you know when I have something worth looking at (or when I get stuck).

hegemonic replied:
We're now using `@microsoft-api-extractor-model`.

We could probably use it a bit more heavily in the future. For example, rather than using a regexp to extract the namespace from the UID, we could first walk the model and create a map of UIDs to nodes; then, to find the namespace for a UID, we could get the node for that UID and walk up its ancestors until we find a namespace node. That might be more robust than the regexp approach. For now, though, the regexp seems to get the job done.

hegemonic commented:
FYI, I haven't forgotten about this PR! I hit a snag, though—when I tested with a client library that supports multiple API versions, I realized that the result was super-confusing. For example, if a client library supports three API versions, you'd end up with a TOC that looks like this:

- `MyServiceConfiguration`
- `MyServiceConfiguration`
- `MyServiceConfiguration`
- `MyServiceRequest`
- `MyServiceRequest`
- `MyServiceRequest`
- etc.

I got that mostly fixed by organizing everything into separate subsections for each version, like so:

- `v2`
    - `MyServiceConfiguration`
    - `MyServiceRequest`
- `v1`
    - `MyServiceConfiguration`
    - `MyServiceRequest`
- `v1beta`
    - `MyServiceConfiguration`
    - `MyServiceRequest`

But I don't like that, either; a version identifier like `v2` doesn't give you enough context about what you'll get when you expand that section.

So now I'm working on [organizing everything by namespace](https://gist.github.com/hegemonic/cb3afc6dc5a3a8e2128facf7e5f88dbc) in this PR, rather than in a follow-up PR. Time has been a little tight, but I'm making progress. I'll keep you posted!

hegemonic committed:
checkpoint

Sorting by version number is working, and all tests pass.

Next up: Grouping by namespace.

hegemonic committed:
namespaces mostly working

The TOC is now grouped by namespace. This feature works with most packages I've tested.

It doesn't work with googleapis/nodejs-storage, which is a non-GAPIC library. (The fact that it's non-GAPIC might not be the cause of the issue; need to investigate further.)

hegemonic committed:
namespaces fully working

I've confirmed that this code works as expected with these client libraries:

- Cloud Functions (has multiple API versions)
- Cloud Gaming (has multiple API versions and multiple clients exported at top level)
- Cloud Storage (non-GAPIC)

hegemonic commented:
@dansaadati This is finally ready for another look. We now group items in each category (Classes, Interfaces, Enums) by namespace. I tested this PR with several client libraries; here are the results:

+ [Cloud Functions](https://gist.github.com/hegemonic/900910c9ecabc634390c3335f56e83ce)
+ [Cloud Game Servers](https://gist.github.com/hegemonic/ccc6a5c760260ecc6f6089215f9cfa6b)
+ [Cloud Storage](https://gist.github.com/hegemonic/6ad17c552bfa0e155be6e9dcd7510fa4)

If there are others that I should test as well, please let me know, and I'll try them out.

dansaadati commented on lib/processors/organize-toc.mjs:
```diff
@@ -0,0 +1,250 @@
+      // Remove ` (Class)` from titles like `FooBar (Class)`.
+      item.name = item.name.replace(TITLE_KIND_REGEXP, '');
+
+      // The package's top-level exported classes get special treatment. To
```
I'm a bit unsure on this behavior as we'd only show the newest version's client class at the top-level, and the other versions would be under the `Classes` grouping. I think there's a couple of options we have here:

1. Show all of the top-level classes and add a version to distinguish them e.g. (`ImageAnnotatorClient (v2)` `ImageAnnotatorClient (v1)`).
2. Group top-level classes by version as well, like: 
``` 
  - name: v2
    items:
      - name: ImageAnnotatorClient
        uid: '@google-cloud/vision!v1p4beta1.ImageAnnotatorClient:class'
  - name: v1
    items:
      - name: ImageAnnotatorClient
        uid: '@google-cloud/vision!v1p4beta1.ImageAnnotatorClient:class'
```

What do you think?

hegemonic replied:
Sounds good. I went with option 1.

dansaadati commented on lib/processors/organize-toc.mjs:
```diff
@@ -0,0 +1,250 @@
+      }
+      // Don't mess with items that have UIDs like `@google-cloud/foo!`, or a
+      // kind that we don't recognize.
+      else if (!kind || !Object.prototype.hasOwnProperty.call(this, kind)) {
```
If I'm understanding correctly, this is not expected to happen, right? I think this behavior is fine, but I think it'd be helpful if we log cases when we see this.

hegemonic replied:
Items with UIDs like `@google-cloud/foo!` are normal and expected. Items with unrecognized kinds are not.

Updated to log a warning only for items with unrecognized kinds.

dansaadati reviewed (commented):
Awesome work! Left a few items to look at, but overall looks great.

hegemonic commented on lib/processors/organize-toc.mjs:
```diff
@@ -0,0 +1,250 @@
+        items: [],
+      };
+      // Add namespaced items to the top of the list.
+      for (const namespace of Object.keys(this[kind].namespaces).sort(
```
I _think_ your goal is to put the stuff for the current package first, followed by all the common protos. To that end, I added a sorter that puts namespaces in this order:

- Bare version identifiers
- Other namespaces that belong to the current package
- All other namespaces

For example, here's how it sorts the namespaces for Cloud Functions classes:

```js
[
  'v1',
  'v2alpha',
  'v2beta',
  'protos.google.cloud.functions.v1',
  'protos.google.cloud.functions.v2',
  'protos.google.cloud.functions.v2alpha',
  'protos.google.cloud.functions.v2beta',
  'protos.google.api',
  'protos.google.iam.v1',
  'protos.google.longrunning',
  'protos.google.protobuf',
  'protos.google.rpc',
  'protos.google.type'
]
```

Hope that works! If not, let me know.

hegemonic replied:
(I meant to say that my personal preference is to alphabetize this sort of list. I think that makes it easier for readers to understand how the list is organized. Happy to try it this way, though! We can always change it later.)

dansaadati replied:
I think that order looks great, that works!

hegemonic committed:
address comments

hegemonic commented:
@dansaadati Okay, hopefully we're just about there. Updated examples are here:

+ [Cloud Functions](https://gist.github.com/hegemonic/900910c9ecabc634390c3335f56e83ce)
+ [Cloud Game Servers](https://gist.github.com/hegemonic/ccc6a5c760260ecc6f6089215f9cfa6b)
+ [Cloud Storage](https://gist.github.com/hegemonic/6ad17c552bfa0e155be6e9dcd7510fa4)

And [here's the diff from your last review](https://github.com/googleapis/nodejs-cloud-rad/pull/77/files/8a52c9073f56781db51225b2349b4215bcd9af25..71a222aa98e7006778e86c2fc431caee0188ae53).

One issue I noticed: For `@google-cloud/game-servers`, the tool isn't able to sort namespaces so that the package's namespaces come first. That's because the package name doesn't match the proto namespace, which is `protos.google.cloud.gaming`. I didn't see a good solution for this issue, but let me know if you think of something.

hegemonic commented:
@dansaadati Any thoughts on the CI failure? It doesn't appear to be related to this PR. The tests pass on my Mac with Node.js 18.16.1.

dansaadati commented:
> @dansaadati Any thoughts on the CI failure? It doesn't appear to be related to this PR. The tests pass on my Mac with Node.js 18.16.1.

Thanks for pointing it out!! I have a fix at https://github.com/googleapis/nodejs-cloud-rad/pull/93

dansaadati commented:
> @dansaadati Okay, hopefully we're just about there. Updated examples are here:
> 
> * [Cloud Functions](https://gist.github.com/hegemonic/900910c9ecabc634390c3335f56e83ce)
> * [Cloud Game Servers](https://gist.github.com/hegemonic/ccc6a5c760260ecc6f6089215f9cfa6b)
> * [Cloud Storage](https://gist.github.com/hegemonic/6ad17c552bfa0e155be6e9dcd7510fa4)
> 
> And [here's the diff from your last review](https://github.com/googleapis/nodejs-cloud-rad/pull/77/files/8a52c9073f56781db51225b2349b4215bcd9af25..71a222aa98e7006778e86c2fc431caee0188ae53).
> 
> One issue I noticed: For `@google-cloud/game-servers`, the tool isn't able to sort namespaces so that the package's namespaces come first. That's because the package name doesn't match the proto namespace, which is `protos.google.cloud.gaming`. I didn't see a good solution for this issue, but let me know if you think of something.

That's okay for now -- thanks for updating the sorting!

dansaadati reviewed (approved):
Amazing stuff, thanks so much for taking this on! Really impressed by this (along with your refactor).

hegemonic committed:
fix: delete unintentionally committed file

dansaadati commented:
Please rebase, CI should be green now!

hegemonic committed:
chore: update snapshots; increase snapshot timeout

hegemonic commented:
> Please rebase, CI should be green now!

@dansaadati Done! Can you trigger OwlBot? After that, I think we're good to go.

hegemonic commented:
@dansaadati i can haz merge?

hegemonic committed:
Merge branch 'main' into toc

dansaadati commented:
> @dansaadati i can haz merge?

yes! 🫡

dansaadati commented:
Merging for @hegemonic .
